jbpm-case-mgmt/jbpm-case-mgmt-impl/src/test/java/org/jbpm/casemgmt/impl/CaseServiceImplTest.java

```diff
@@ -3236,4 +3238,100 @@ public void testGetCaseFileInstanceForNotExistingCase() {
         assertThatExceptionOfType(CaseNotFoundException.class)
                 .isThrownBy(() -> caseService.getCaseFileInstance("nonexisting"));
     }
+    
+    @Test
+    public void testStartCaseForMultiInstanceStagesTriggerAdHocFragment() {
+        String caseId = caseService.startCase(deploymentUnit.getIdentifier(), MULTI_STAGE_CASE_P_ID);
+        assertNotNull(caseId);
+        assertEquals(FIRST_CASE_ID, caseId);
+        try {
+            CaseInstance cInstance = caseService.getCaseInstance(caseId);
+            assertNotNull(cInstance);
+            assertEquals(deploymentUnit.getIdentifier(), cInstance.getDeploymentId());
+            
+            Map<String, Object> data = new HashMap<>();
+            data.put("type", "basic");
+            caseService.triggerAdHocFragment(caseId, "Stage #{type}", data);
```

Comment:
Just a question on how this work. So if we don't have NodeInstance created yet, we have to call it as if directly from case definition, i.e. without variables resolved. Once the instance of the stage is created, variables are resolved, right? Just thinking if it makes sense/is possible to do variable resolution even when triggering adHoc fragment, although I see that it can cause troubles since once it would be Stage basic, once Stage advanced, but it would still be the same Stage being invoked. So just an idea...

Reply:
in general the stage needs to be created based on the definition - adhoc subprocess and if that has the name with expression it needs to be used as the way to created it but at the time it's created it will resolve the variable to make sure that stage instance has the correct name instead of expression.